### DISCUSSION TITLE
"Error 12: NotImplemented" received for "dns resolve" command

### DISCUSSION THREAD

**Discussion by clearasu** [openthread/openthread#8555] [Q&A]
I have a very simple thread setup, a border router and a thread accessory/client (nRF52840 board running openthread).

The "dns resolve" on the client always returns "Error 12: NotImplemented". On the border router, tcpdump is running and no packets are seen at all.  However, the ping traffic and even srp registration work fine. 

It looks to me that the "Error 12: NotImplemented" is not from server's response as the border router does not receive any dns packets. It's more like from openthread internally.

The border router recently just moved from NCP to RCP. Would this be relevant?

> srp client server
[fd76:e379:fd30:0:dbeb:d138:a1a2:5ebe]:63755
Done
> dns config
Server: [fd76:e379:fd30:0:dbeb:d138:a1a2:5ebe]:53
ResponseTimeout: 6000 ms
MaxTxAttempts: 3
RecursionDesired: yes
Done
> srp client host name ot-host
Done
> srp client host address fdcf:57e9:c010:0:db11:6c36:b50f:e879
Done
> srp client service add ot-service _ipps._tcp 12345
Done
> srp client service
instance:"ot-service", name:"_ipps._tcp", state:Registered, port:12345, priority:0, weight:0
Done
> dns resolve google.com
DNS response for google.com. - 
Error 12: NotImplemented

**Comment by jwhui** [comment]
I'd like to get more visibility to diagnose your issue. Is it possible for you to provide the following:
- DEBG-level logs from `otbr-agent` on the border router?
- Packet capture from a sniffer for analysis in Wireshark?

**Reply by abtink** [reply]
This is very likely an error response coming from server (name resolver). There is no code path on `Dns::Clinet` itself that would return `kNotImplemnetedError`. The errors from server are mapped to `otError`:
https://github.com/openthread/openthread/blob/50ed30fcd927c2be5793109688128faad5e3185a/include/openthread/dns_client.h#L166-L182


Some suggestions/questions:
- `tcpdump` on host netif will show all traffic that OT stack passes to the host netif but if an IPv6 message is directly processed by OT stack itself it may not be passed to host netif . In this case, if we enable DNS resolver feature on `otbr` I think OT stack will process the query message directly (so that may be why it may look it is not seen on server in tcpdump output).
- Do you have "discovery proxy" setup and running? Do you use `otbr` implementation?
- I expect query for `google.com` would go to discovery proxy and which would then generate the response.
- I think this error may be coming from discovery proxy.

**Reply by jwhui** [reply]
Ah, yes, I missed that you were trying to resolve `google.com`. OTBR doesn't yet implement recursive resolver functionality. The Discovery Proxy only supports mDNS currently.

For reference, the entry point for the discovery proxy occurs here:
https://github.com/openthread/openthread/blob/1641eced0d0b19828899e0f75fc8ce7a25664cab/src/core/net/dnssd_server.cpp#L176-L183

And the current OTBR implementation implements the callback here:
https://github.com/openthread/ot-br-posix/blob/fbbb62fafce55cdd8c0fa2352b2e27d30c2d2192/src/sdp_proxy/discovery_proxy.cpp#L111-L129

As you can see, OTBR will only attempt to query using mDNS.

**Reply by clearasu** [reply]
Our border router does not use otbr implementation, and yes we have discovery proxy setup and running which implemented recursive resolver functionality.

**Reply by jwhui** [reply]
@clearasu , it would be great if you can provide DEBG-level logging and/or packet sniffer capture to provide more visibility into the issue. Thanks.
